runtime/src/mock_channel.rs

```diff
@@ -0,0 +1,181 @@
+                (os_send, Bytes::from(data))
+            } else {
+                return Ok(());
+            }
+        };
+
+        // Resolve the waiter.
+        os_send.send(data).map_err(|_| Error::SendFailed)?;
+        Ok(())
+    }
+}
+
+// A mock stream that implements the Stream trait.
+pub struct ByteStream {
+    channel: Arc<Mutex<ByteChannel>>,
+}
+
+impl Stream for ByteStream {
+    // Blocks until the buffer has enough bytes to fill `buf`.
+    // Does not hold the lock unless it needs to.
+    async fn recv(&mut self, buf: &mut [u8]) -> Result<(), Error> {
+        let os_recv = {
+            let mut channel = self.channel.lock().unwrap();
+
+            // If the message is fully available in the buffer,
+            // drain the value into buf and return.
+            if channel.buffer.len() >= buf.len() {
+                let b: Vec<u8> = channel.buffer.drain(0..buf.len()).collect();
+                buf.copy_from_slice(&b);
+                return Ok(());
+            }
+
+            // Otherwise, create a oneshot receiver and store it in the channel.
+            assert!(channel.waiter.is_none());
+            let (os_send, os_recv) = oneshot::channel();
+            channel.waiter = Some((buf.len(), os_send));
+            os_recv
+        };
+
+        // Wait for the oneshot receiver to be resolved.
+        let data = os_recv.await.map_err(|_| Error::RecvFailed)?;
+        buf.copy_from_slice(&data);
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::executor::block_on;
+
+    #[test]
+    fn test_send_recv() {
+        let (mut sink, mut stream) = new();
+
+        let data = b"hello world";
+        let mut buf = vec![0; data.len()];
+
+        block_on(async {
+            sink.send(data).await.unwrap();
+            stream.recv(&mut buf).await.unwrap();
+        });
+
+        assert_eq!(buf, data);
+    }
+
+    #[test]
+    fn test_send_recv_partial_multiple() {
+        let (mut sink, mut stream) = new();
+
+        let data1 = b"hello";
+        let data2 = b"world";
+        let mut buf1 = vec![0; data1.len()];
+        let mut buf2 = vec![0; data2.len()];
+
+        block_on(async {
+            sink.send(data1).await.unwrap();
+            sink.send(data2).await.unwrap();
+            stream.recv(&mut buf1[0..3]).await.unwrap();
+            stream.recv(&mut buf1[3..]).await.unwrap();
+            stream.recv(&mut buf2).await.unwrap();
+        });
+
+        assert_eq!(buf1, data1);
+        assert_eq!(buf2, data2);
+    }
+
+    #[test]
+    fn test_send_recv_async() {
+        let (mut sink, mut stream) = new();
+
+        let data = b"hello world";
+        let mut buf = vec![0; data.len()];
+
+        block_on(async {
+            futures::try_join!(
+                sink.send(data),
+                stream.recv(&mut buf),
+            )
+            .unwrap();
+        });
+
+        assert_eq!(buf, data);
+    }
+
+    /*
```

Comment:
I wouldn't worry about testing these for now. Because of our approach, forcing an error on the oneshot would require mutating internal variables (which I tend to draw a line at when writing tests).

Reply:
I suppose the alternative would be to add an explicit "fail" function somewhere (in case someone wants to test that explicitly) but feel that should be done later tbh.